File: app/src/main/java/com/mikhailovskii/trakttv/ui/adapter/MoviesAdapter.java
package com.mikhailovskii.trakttv.ui.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.mikhailovskii.trakttv.R;
import com.mikhailovskii.trakttv.data.entity.Movie;

import java.util.ArrayList;
import java.util.List;

public class MoviesAdapter extends RecyclerView.Adapter<MoviesAdapter.ViewHolder> {

    private List<Movie> moviesList = new ArrayList<>();
    private OnItemClickListener onItemClickListener;

    public MoviesAdapter(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.movie_element, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int position) {
        Movie movie = moviesList.get(position);

        Glide.with(viewHolder.itemView.getContext())
                .load(movie.getIconUrl())
                .into(viewHolder.mIconImageView);

        viewHolder.mMovieNameTextView.setText(movie.getName());
        viewHolder.mYearTextView.setText("Year: " + movie.getYear());
        viewHolder.mWatchersTextView.setText("Watchers: " + movie.getWatchers());

        viewHolder.itemView.setOnClickListener(v -> {
            if (viewHolder.getAdapterPosition() != RecyclerView.NO_POSITION) {
                onItemClickListener.onItemClicked(viewHolder.getAdapterPosition(), movie);
            }
        });

        viewHolder.itemView.setOnLongClickListener(v -> {
            if (viewHolder.getAdapterPosition() != RecyclerView.NO_POSITION) {
                onItemClickListener.onItemLongClick(viewHolder.getAdapterPosition(), movie);
            }
            return false;
        });
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemCount() {
        return moviesList.size();
    }

    public void setData(List<Movie> movies) {
        moviesList.clear();
        moviesList.addAll(movies);
        notifyDataSetChanged();
    }

    public interface OnItemClickListener {
        void onItemClicked(int position, Movie item);

        void onItemLongClick(int position, Movie item);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        ImageView mIconImageView;
        TextView mMovieNameTextView;
        TextView mYearTextView;
        TextView mWatchersTextView;

        ViewHolder(@NonNull View itemView) {
            super(itemView);

            mIconImageView = itemView.findViewById(R.id.icon_imageview);
            mMovieNameTextView = itemView.findViewById(R.id.moviename_textview);
            mYearTextView = itemView.findViewById(R.id.year_textview);
            mWatchersTextView = itemView.findViewById(R.id.watchers);
        }

    }

}

File: app/src/main/java/com/mikhailovskii/trakttv/ui/adapter/MoviesAdapter.kt
package com.mikhailovskii.trakttv.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.mikhailovskii.trakttv.R
import com.mikhailovskii.trakttv.data.entity.Movie
import kotlinx.android.synthetic.main.movie_element.view.*
import java.util.*

class MoviesAdapter(
        private val onItemClickListener: OnItemClickListener
) : RecyclerView.Adapter<MoviesAdapter.ViewHolder>() {

    private val movieList = ArrayList<Movie>()

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.movie_element, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.bindData(movieList[position], onItemClickListener)
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemCount(): Int {
        return movieList.size
    }

    fun setData(movies: List<Movie>) {
        movieList.clear()
        movieList.addAll(movies)
        notifyDataSetChanged()
    }

    interface OnItemClickListener {

        fun onItemClicked(position: Int, item: Movie)

        fun onItemLongClick(position: Int, item: Movie)

    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bindData(movie: Movie, onItemClickListener: OnItemClickListener) {
            Glide.with(itemView.context)
                    .load(movie.iconUrl)
                    .into(itemView.icon_image)

            itemView.tv_moviename.text = movie.name
            itemView.tv_year.text = itemView.context.resources.getString(R.string.year, movie.year)
            itemView.tv_watchers.text = itemView.context.resources.getString(R.string.watchers, movie.watchers)

            itemView.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    onItemClickListener.onItemClicked(adapterPosition, movie)
                }
            }

            itemView.setOnLongClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    onItemClickListener.onItemLongClick(adapterPosition, movie)
                }
                true
            }
        }

    }

}
